File: 338-counting-bits/counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        DP = [0] * (n + 1)
        offset = 1

        for i in range(n + 1):
            if i == 0:
                DP[i] = 0
                continue

            if offset * 2 == i:
                offset = i

            DP[i] = 1 + DP[i - offset]
        
        return DP

            
            

File: 338-counting-bits/counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ones(n + 1);
        int powTwo = 1;
        int count = 0;
        for(int i = 0; i <= n; i++){
            if(i == 0){
                ones[i] == 0;
                continue;
            }
            else if(i == powTwo){
                powTwo *= 2;
                count = 0;
            }
            ones[i] = 1 + ones[count];
            count++;
        }

        return ones;
    }
};